kratos/python_scripts/analysis_stage.py

```diff
@@ -1,305 +1,384 @@
+            self.InitializeSolutionStep()
+            self._GetSolver().Predict()
+            is_converged = self._GetSolver().SolveSolutionStep()
+            self.__CheckIfSolveSolutionStepReturnsAValue(is_converged)
+            self.FinalizeSolutionStep()
+            self.OutputSolutionStep()
+
+    def Initialize(self):
+        """This function initializes the AnalysisStage
+        Usage: It is designed to be called ONCE, BEFORE the execution of the solution-loop
+        This function has to be implemented in deriving classes!
+        """
+        self._SetupGeometryModelAndModelParts()
+
+        self._GetSolver().ImportModelPart()
+        self._GetSolver().PrepareModelPart()
+        self._GetSolver().AddDofs()
+
+        self.ModifyInitialProperties()
+        self.ModifyInitialGeometry()
+
+        ##here we initialize user-provided processes
+        self.__CreateListOfProcesses() # has to be done after importing and preparing the ModelPart
+        for process in self._GetListOfProcesses():
+            process.ExecuteInitialize()
+
+        self._GetSolver().Initialize()
+        self.Check()
+
+        self.ModifyAfterSolverInitialize()
+
+        for process in self._GetListOfProcesses():
+            process.ExecuteBeforeSolutionLoop()
+
+        ## Stepping and time settings
+        self.end_time = self.project_parameters["problem_data"]["end_time"].GetDouble()
+
+        if self._GetSolver().GetComputingModelPart().ProcessInfo[KratosMultiphysics.IS_RESTARTED]:
+            self.time = self._GetSolver().GetComputingModelPart().ProcessInfo[KratosMultiphysics.TIME]
+        else:
+            self.time = self.project_parameters["problem_data"]["start_time"].GetDouble()
+
+        ## If the echo level is high enough, print the complete list of settings used to run the simualtion
+        if self.echo_level > 1:
+            with open("ProjectParametersOutput.json", 'w') as parameter_output_file:
+                parameter_output_file.write(self.project_parameters.PrettyPrintJsonString())
+
+        KratosMultiphysics.Logger.PrintInfo(self._GetSimulationName(), "Analysis -START- ")
+
+    def Finalize(self):
+        """This function finalizes the AnalysisStage
+        Usage: It is designed to be called ONCE, AFTER the execution of the solution-loop
+        """
+        for process in self._GetListOfProcesses():
+            process.ExecuteFinalize()
+
+        self._GetSolver().Finalize()
+
+        KratosMultiphysics.Logger.PrintInfo(self._GetSimulationName(), "Analysis -END- ")
+
+    def InitializeSolutionStep(self):
+        """This function performs all the required operations that should be executed
+        (for each step) BEFORE solving the solution step.
+        """
+        self.PrintAnalysisStageProgressInformation()
+
+        # updates, modifies and maps within the models.
+        for modeler in self._GetListOfModelers():
+            modeler.UpdateModelInitializeSolutionStep(self.model)

```

Comment:
you miss the debugging echos...also this should be included in a function

Reply:
The modeler has the mEchoLevel with which the output of each modeler is controlled. Are you thinking of something additional to that?

Reply:
ok